ZEPXXX-priority-election.md

```diff
@@ -0,0 +1,231 @@
+Previously, on heartbeat timeout a follower immediately starts an election by sending a poll request.
+The heartbeat timeout was also randomized in order to prevent multiple nodes starting the election at the same time.
+With the priority election we do not use a randomized heartbeat timeout.
+Instead we use the configured electionTimeout.
+On heartbeat timeout, the follower implements the following algorithm.
+
+
+```java
+onHeartbeatTimeout() {
+    if(nodePriority >= targetPriority) {
+        sendPollRequest();
+    } else {
+        // reduce priority in the next heartbeatTimeout
+        heartbeatTimer = schedule(electionTimeout, () -> {
+            targetPriority = targetPriority - priorityDecay;
+            onHeartbeatTimeout();
+        })
+    }
+}
+```
+
+
+### Assigning priority
+
+The node with highest priority is called the primary for the partition.
+If a node is primary for two partitions a and b, then the next priority for each partitions should be assigned to different nodes.
+
+Here is an example how we assign the priorities. Each node gets a priority value for each partition.
+
+| node\partition | 1 | 2 | 3 | 4 | 5 | 6 |
+| --------       |:------:| -----:|:-----:| -----:|:------:| -----:|
+| 0              | 3 | 1 | 2 | 3 | 2 | 1 |
+| 1              | 2 | 3 | 1 | 1 | 3 | 2 |
+| 2              | 1 | 2 | 3 | 2 | 1 | 3 |
+
+Node 0 is the primary for partitions 1 and 4. If node 1 has the next priority for both partitions, then when node 0 dies node 1 will become the leader for both partitions.
+This is not optimal as the leadership is not balanced. Hence it is preferable to make different nodes as the secondary priorities. As in the above example, node 1 is the secondary for partition 1, while node 2 is the secondary for partition 4.
+
+Here is how the priority is assigned.
+The primary node get the highest priority which is equal to the replicationFactor.
+
+```java
+primaryMember.priority = replicationFactor;
+if(partition/clustersize  % 2 == 0) {
+    nextPriority = 1;
+    foreach (member : nonPrimaryPartitionMembers) {
+        member.priority = nextPriority;
+        nextPriority++;
+    }
+} else {
+    nextPriority = replicationFactor - 1
+    foreach (member : nonPrimaryPartitionMembers) {
+        member.priority = nextPriority;
+        nextPriority--;
+    }
+}
+```
+
+We propose to have the priorities in the range of 1 to `replicationFactor` and set `priorityDecay` to 1.
+Since the system does not allow dynamic change of the configuration, we can calculate and assign the priorities when the partition distribution is generated.
+The priorities will be part of `PartitionMetadata`.
+
+### Possible issues and corner cases
+1. If even after multiple rounds of election no new leader is elected, it is possible that all nodes starts election at the same time and enters an election loop. <To be verified>
+2. In normal cases, the election can take upto ((quorum + 1) * electionTimeout) until a new leader is elected.
```

Comment:
In avg it will take longer to elect an new leader right?

Reply:
Yes.